core/sr-primitives/src/weights.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Primitives for transaction weighting.
+//!
+//! Each dispatch function withing `decl_module!` can now have an optional
+//! `#[weight = $x]` attribute. $x can be any object that implements the
+//! [`Weighable`] trait. By default, All transactions are annotated by
+//! `#[weight = TransactionWeight::default()]`.
+//!
+//! Note that the decl_module macro _cannot_ enforce this and will simply fail
+//! if an invalid struct is passed in.
+
+/// The final type that each `#[weight = $x:expr]`'s
+/// expression must evaluate to.
+pub type Weight = u32;
+
+/// A `Call` enum (aka transaction) that can be weighted using the custom weight attribute of the
+/// its dispatchable functions. Is implemented by default in the `decl_module!`.
+///
+/// Both the outer Call enum and the per-module individual ones will implement this.
+/// The outer enum simply calls the inner ones based on call type.
+pub trait Weighable {
+	/// Return the weight of this call.
+	/// The `len` argument is the number of bytes in the transaction/call.
+	fn weight(&self, len: usize) -> Weight;
+}
+
+/// Default weight calculator.
+/// This is tailored for the Polkadot use case. Users may replace it with anything.
```

Comment:
Would drop Polkadot in the description and give more information how the weight is calculated.